src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -0,0 +1,236 @@
+        if isinstance(domain, SymbolicDomain):
+            return domain
+        if isinstance(domain, itir.FunCall) and domain.fun == im.ref("cartesian_domain"):
+            return SymbolicDomain.from_expr(domain)
+        raise TypeError("domain must either be a FunCall or a SymbolicDomain.")
+
+    @staticmethod
+    def _merge_domains(
+        original_domains: Dict[str, SymbolicDomain], new_domains: Dict[str, SymbolicDomain]
+    ) -> Dict[str, SymbolicDomain]:
+        for key, value in new_domains.items():
+            if key in original_domains:
+                original_domains[key] = domain_union([original_domains[key], value])
+            else:
+                original_domains[key] = value
+
+        return {
+            key: domain_union(value) if isinstance(value, list) else value
+            for key, value in original_domains.items()
+        }
+
+    @classmethod
+    def _translate_domain(
+        cls,
+        symbolic_domain: SymbolicDomain,
+        shift: Tuple[itir.OffsetLiteral, itir.OffsetLiteral],
+        offset_provider: Dict[str, Dimension],
+    ) -> SymbolicDomain:
+        dims = list(symbolic_domain.ranges.keys())
+        new_ranges = {dim: symbolic_domain.ranges[dim] for dim in dims}
+        if shift:
+            off, val = shift
+            nbt_provider = offset_provider[off.value]
+            if isinstance(nbt_provider, common.Dimension):
+                current_dim = nbt_provider
+                # cartesian offset
+                new_ranges[current_dim] = SymbolicRange.translate(
+                    symbolic_domain.ranges[current_dim], val.value
+                )
+            elif isinstance(nbt_provider, common.Connectivity):
+                # unstructured shift
+                # TODO: move to initialization
+                horizontal_sizes = _max_domain_sizes_by_location_type(offset_provider)
+
+                nbt_provider = offset_provider[off.value]
+                old_dim = nbt_provider.origin_axis
+                new_dim = nbt_provider.neighbor_axis
+
+                assert new_dim not in new_ranges or old_dim == new_dim
+
+                # TODO(tehrengruber): symbolic sizes for ICON?
+                new_range = SymbolicRange(
+                    im.literal("0", itir.INTEGER_INDEX_BUILTIN),
+                    im.literal(
+                        str(horizontal_sizes[new_dim.value]), itir.INTEGER_INDEX_BUILTIN
+                    ),
+                )
+                new_ranges = dict(
+                    (dim, range_) if dim != old_dim else (new_dim, new_range)
+                    for dim, range_ in new_ranges.items()
+                )
+            else:
+                raise AssertionError()
+
+        return SymbolicDomain(symbolic_domain.grid_type, new_ranges)
+
+    @classmethod
+    def infer_as_fieldop(
+        cls,
+        applied_fieldop: itir.FunCall,
+        input_domain: SymbolicDomain,
```

Comment:
```suggestion
        input_domain: SymbolicDomain | ir.FunCall,
```
and then below you can do the canonicalization
```
if isinstance(input_domain, ir.FunCall):
  input_domain = SymbolicDomain.from_expr(input_domain)
```
The `_get_symbolic_domain` is then not needed anymore.